utils/s3/client.cc

```diff
@@ -250,30 +243,61 @@ storage_io_error map_s3_client_exception(std::exception_ptr ex) {
         auto e = std::current_exception();
         return {EIO, format("S3 error ({})", e)};
     }
-
 }
 
-future<> client::do_make_request(group_client& gc, http::request req, http::experimental::client::reply_handler handle, std::optional<http::reply::status_type> expected_opt, seastar::abort_source* as) {
+future<> client::do_retryable_request(group_client& gc, http::request req, http::experimental::client::reply_handler handler, seastar::abort_source* as) const {
     // TODO: the http client does not check abort status on entry, and if
     // we're already aborted when we get here we will paradoxally not be
     // interrupted, because no registration etc will be done. So do a quick
     // preemptive check already.
     if (as && as->abort_requested()) {
-        return make_exception_future<>(as->abort_requested_exception_ptr());
-    }
-    auto expected = expected_opt.value_or(http::reply::status_type::ok);
-    return (as
-        ? gc.http.make_request(std::move(req), std::move(handle), *as, expected)
-        : gc.http.make_request(std::move(req), std::move(handle), expected)
-        ).handle_exception([] (auto ex) {
-            return map_s3_client_exception(std::move(ex));
-        });
+        co_await coroutine::return_exception_ptr(as->abort_requested_exception_ptr());
+    }
+    uint32_t retries = 0;
+    std::exception_ptr e;
+    aws::aws_exception request_ex{aws::aws_error{aws::aws_error_type::OK, aws::retryable::yes}};
+    while (true) {
+        try {
+            e = {};
+            co_return co_await (as ? gc.http.make_request(req, handler, *as, std::nullopt) : gc.http.make_request(req, handler, std::nullopt));
```

Comment:
How does it work if `as == nullptr`?

Reply:
> How does it work if `as == nullptr`?

It uses [make_request](https://github.com/scylladb/seastar/blob/fb6c969ccf2f7735fdb43e3f9beac8b68711e29a/include/seastar/http/client.hh#L286) without the abort source.
Isn't it?